Releases > 0.3.5 fail to work with docker 1.8.2 client unless the base image is already on the host

When upgrading the version of the docker-maven-plugin, I ran into an issue that I narrowed down to changes in 0.3.6 (I suspect the docker-client bump from 3.3.1 to 3.3.4).

In order to reproduce the error you have to be using a docker version > 1.7 (that supports the v2 protocol) and be pulling images from a repo which also supports the /v2 api AND not have the base image already on the docker host.

When those conditions are met there appears to be some sort of race conditions, that cause the plugin's class path to miss guava:

[ERROR] Failed to execute goal com.spotify:docker-maven-plugin:0.3.6:build (default) on project testing-docker: Execution default of goal com.spotify:docker-maven-plugin:0.3.6:build failed: A required class was missing while executing com.spotify:docker-maven-plugin:0.3.6:build: com/google/common/base/MoreObjects
[ERROR] -----------------------------------------------------
[ERROR] realm =    plugin>com.spotify:docker-maven-plugin:0.3.6
[ERROR] strategy = org.codehaus.plexus.classworlds.strategy.SelfFirstStrategy
[ERROR] urls[0] = file:/c:/dev/Maven/com/spotify/docker-maven-plugin/0.3.6/docker-maven-plugin-0.3.6.jar
[ERROR] urls[1] = file:/c:/dev/Maven/com/spotify/docker-client/3.3.1/docker-client-3.3.1-shaded.jar
[ERROR] urls[2] = file:/c:/dev/Maven/org/slf4j/slf4j-api/1.7.12/slf4j-api-1.7.12.jar
[ERROR] urls[3] = file:/c:/dev/Maven/com/google/guava/guava/14.0.1/guava-14.0.1.jar
[ERROR] urls[4] = file:/c:/dev/Maven/com/fasterxml/jackson/jaxrs/jackson-jaxrs-json-provider/2.6.0/jackson-jaxrs-json-provider-2.6.0.jar
[ERROR] urls[5] = file:/c:/dev/Maven/com/fasterxml/jackson/jaxrs/jackson-jaxrs-base/2.6.0/jackson-jaxrs-base-2.6.0.jar
[ERROR] urls[6] = file:/c:/dev/Maven/com/fasterxml/jackson/module/jackson-module-jaxb-annotations/2.6.0/jackson-module-jaxb-annotations-2.6.0.jar
[ERROR] urls[7] = file:/c:/dev/Maven/com/fasterxml/jackson/datatype/jackson-datatype-guava/2.6.0/jackson-datatype-guava-2.6.0.jar
[ERROR] urls[8] = file:/c:/dev/Maven/org/glassfish/jersey/core/jersey-client/2.19/jersey-client-2.19.jar
[ERROR] urls[9] = file:/c:/dev/Maven/javax/ws/rs/javax.ws.rs-api/2.0.1/javax.ws.rs-api-2.0.1.jar
[ERROR] urls[10] = file:/c:/dev/Maven/org/glassfish/jersey/core/jersey-common/2.19/jersey-common-2.19.jar
[ERROR] urls[11] = file:/c:/dev/Maven/javax/annotation/javax.annotation-api/1.2/javax.annotation-api-1.2.jar
[ERROR] urls[12] = file:/c:/dev/Maven/org/glassfish/jersey/bundles/repackaged/jersey-guava/2.19/jersey-guava-2.19.jar
[ERROR] urls[13] = file:/c:/dev/Maven/org/glassfish/hk2/osgi-resource-locator/1.0.1/osgi-resource-locator-1.0.1.jar
[ERROR] urls[14] = file:/c:/dev/Maven/org/glassfish/hk2/hk2-api/2.4.0-b25/hk2-api-2.4.0-b25.jar
[ERROR] urls[15] = file:/c:/dev/Maven/org/glassfish/hk2/hk2-utils/2.4.0-b25/hk2-utils-2.4.0-b25.jar
[ERROR] urls[16] = file:/c:/dev/Maven/org/glassfish/hk2/external/aopalliance-repackaged/2.4.0-b25/aopalliance-repackaged-2.4.0-b25.jar
[ERROR] urls[17] = file:/c:/dev/Maven/org/glassfish/hk2/external/javax.inject/2.4.0-b25/javax.inject-2.4.0-b25.jar
[ERROR] urls[18] = file:/c:/dev/Maven/org/glassfish/hk2/hk2-locator/2.4.0-b25/hk2-locator-2.4.0-b25.jar
[ERROR] urls[19] = file:/c:/dev/Maven/org/javassist/javassist/3.18.1-GA/javassist-3.18.1-GA.jar
[ERROR] urls[20] = file:/c:/dev/Maven/org/glassfish/jersey/connectors/jersey-apache-connector/2.19/jersey-apache-connector-2.19.jar
[ERROR] urls[21] = file:/c:/dev/Maven/org/glassfish/jersey/media/jersey-media-json-jackson/2.19/jersey-media-json-jackson-2.19.jar
[ERROR] urls[22] = file:/c:/dev/Maven/org/glassfish/jersey/ext/jersey-entity-filtering/2.19/jersey-entity-filtering-2.19.jar
[ERROR] urls[23] = file:/c:/dev/Maven/org/apache/commons/commons-compress/1.9/commons-compress-1.9.jar
[ERROR] urls[24] = file:/c:/dev/Maven/org/apache/httpcomponents/httpclient/4.5/httpclient-4.5.jar
[ERROR] urls[25] = file:/c:/dev/Maven/org/apache/httpcomponents/httpcore/4.4.1/httpcore-4.4.1.jar
[ERROR] urls[26] = file:/c:/dev/Maven/commons-logging/commons-logging/1.2/commons-logging-1.2.jar
[ERROR] urls[27] = file:/c:/dev/Maven/commons-codec/commons-codec/1.9/commons-codec-1.9.jar
[ERROR] urls[28] = file:/c:/dev/Maven/com/github/jnr/jnr-unixsocket/0.8/jnr-unixsocket-0.8.jar
[ERROR] urls[29] = file:/c:/dev/Maven/com/github/jnr/jnr-ffi/2.0.3/jnr-ffi-2.0.3.jar
[ERROR] urls[30] = file:/c:/dev/Maven/com/github/jnr/jffi/1.2.9/jffi-1.2.9.jar
[ERROR] urls[31] = file:/c:/dev/Maven/com/github/jnr/jffi/1.2.9/jffi-1.2.9-native.jar
[ERROR] urls[32] = file:/c:/dev/Maven/org/ow2/asm/asm/5.0.3/asm-5.0.3.jar
[ERROR] urls[33] = file:/c:/dev/Maven/org/ow2/asm/asm-commons/5.0.3/asm-commons-5.0.3.jar
[ERROR] urls[34] = file:/c:/dev/Maven/org/ow2/asm/asm-analysis/5.0.3/asm-analysis-5.0.3.jar
[ERROR] urls[35] = file:/c:/dev/Maven/org/ow2/asm/asm-tree/5.0.3/asm-tree-5.0.3.jar
[ERROR] urls[36] = file:/c:/dev/Maven/org/ow2/asm/asm-util/5.0.3/asm-util-5.0.3.jar
[ERROR] urls[37] = file:/c:/dev/Maven/com/github/jnr/jnr-x86asm/1.0.2/jnr-x86asm-1.0.2.jar
[ERROR] urls[38] = file:/c:/dev/Maven/com/github/jnr/jnr-constants/0.8.7/jnr-constants-0.8.7.jar
[ERROR] urls[39] = file:/c:/dev/Maven/com/github/jnr/jnr-enxio/0.9/jnr-enxio-0.9.jar
[ERROR] urls[40] = file:/c:/dev/Maven/com/github/jnr/jnr-posix/3.0.12/jnr-posix-3.0.12.jar
[ERROR] urls[41] = file:/c:/dev/Maven/commons-lang/commons-lang/2.6/commons-lang-2.6.jar
[ERROR] urls[42] = file:/c:/dev/Maven/org/bouncycastle/bcpkix-jdk15on/1.52/bcpkix-jdk15on-1.52.jar
[ERROR] urls[43] = file:/c:/dev/Maven/org/bouncycastle/bcprov-jdk15on/1.52/bcprov-jdk15on-1.52.jar
[ERROR] urls[44] = file:/c:/dev/Maven/com/typesafe/config/1.2.0/config-1.2.0.jar
[ERROR] urls[45] = file:/c:/dev/Maven/org/eclipse/aether/aether-util/0.9.0.M2/aether-util-0.9.0.M2.jar
[ERROR] urls[46] = file:/c:/dev/Maven/javax/enterprise/cdi-api/1.0/cdi-api-1.0.jar
[ERROR] urls[47] = file:/c:/dev/Maven/javax/annotation/jsr250-api/1.0/jsr250-api-1.0.jar
[ERROR] urls[48] = file:/c:/dev/Maven/javax/inject/javax.inject/1/javax.inject-1.jar
[ERROR] urls[49] = file:/c:/dev/Maven/org/sonatype/sisu/sisu-guice/3.1.0/sisu-guice-3.1.0-no_aop.jar
[ERROR] urls[50] = file:/c:/dev/Maven/aopalliance/aopalliance/1.0/aopalliance-1.0.jar
[ERROR] urls[51] = file:/c:/dev/Maven/org/eclipse/sisu/org.eclipse.sisu.inject/0.0.0.M5/org.eclipse.sisu.inject-0.0.0.M5.jar
[ERROR] urls[52] = file:/c:/dev/Maven/org/codehaus/plexus/plexus-interpolation/1.19/plexus-interpolation-1.19.jar
[ERROR] urls[53] = file:/c:/dev/Maven/org/codehaus/plexus/plexus-utils/3.0.17/plexus-utils-3.0.17.jar
[ERROR] urls[54] = file:/c:/dev/Maven/org/codehaus/plexus/plexus-component-annotations/1.5.5/plexus-component-annotations-1.5.5.jar
[ERROR] urls[55] = file:/c:/dev/Maven/org/sonatype/plexus/plexus-sec-dispatcher/1.3/plexus-sec-dispatcher-1.3.jar
[ERROR] urls[56] = file:/c:/dev/Maven/org/sonatype/plexus/plexus-cipher/1.4/plexus-cipher-1.4.jar
[ERROR] urls[57] = file:/c:/dev/Maven/org/eclipse/jgit/org.eclipse.jgit/3.2.0.201312181205-r/org.eclipse.jgit-3.2.0.201312181205-r.jar
[ERROR] urls[58] = file:/c:/dev/Maven/com/jcraft/jsch/0.1.46/jsch-0.1.46.jar
[ERROR] urls[59] = file:/c:/dev/Maven/com/googlecode/javaewah/JavaEWAH/0.5.6/JavaEWAH-0.5.6.jar
[ERROR] urls[60] = file:/c:/dev/Maven/com/fasterxml/jackson/core/jackson-databind/2.2.3/jackson-databind-2.2.3.jar
[ERROR] urls[61] = file:/c:/dev/Maven/com/fasterxml/jackson/core/jackson-annotations/2.2.3/jackson-annotations-2.2.3.jar
[ERROR] urls[62] = file:/c:/dev/Maven/com/fasterxml/jackson/core/jackson-core/2.2.3/jackson-core-2.2.3.jar
[ERROR] Number of foreign imports: 1
[ERROR] import: Entry[import  from realm ClassRealm[project>com.hello:testing-docker:1.0.0-SNAPSHOT, parent: ClassRealm[maven.api, parent: null]]]


if however the base image is already on the docker host, build works as expected.

--- a/pom.xml
+++ b/pom.xml
@@ -142,12 +142,12 @@
     <dependency>
       <groupId>org.apache.maven</groupId>
       <artifactId>maven-core</artifactId>
-      <version>3.2.1</version>
+      <version>3.3.3</version>
     </dependency>
     <dependency>
       <groupId>org.apache.maven</groupId>
       <artifactId>maven-artifact</artifactId>
-      <version>3.2.1</version>
+      <version>3.3.3</version>
     </dependency>
     <dependency>
       <groupId>org.eclipse.jgit</groupId>
@@ -157,7 +157,7 @@
     <dependency>
       <groupId>org.apache.maven.plugin-testing</groupId>
       <artifactId>maven-plugin-testing-harness</artifactId>
-      <version>3.1.0</version>
+      <version>3.3.0</version>
       <scope>test</scope>
       <exclusions>
         <exclusion>
@@ -174,7 +174,7 @@
     <dependency>
       <groupId>org.apache.maven</groupId>
       <artifactId>maven-compat</artifactId>
-      <version>3.1.1</version>
+      <version>3.3.3</version>
       <scope>test</scope>
     </dependency>
     <dependency>
@@ -209,16 +209,6 @@
     </dependency>
   </dependencies>
 
-  <dependencyManagement>
-    <dependencies>
-      <dependency>
-        <groupId>com.google.guava</groupId>
-        <artifactId>guava</artifactId>
-        <version>14.0.1</version>
-      </dependency>
-    </dependencies>
-  </dependencyManagement>
-
   <build>
     <testResources>
       <testResource>
@@ -240,7 +230,7 @@
       <plugin>
         <groupId>org.apache.maven.plugins</groupId>
         <artifactId>maven-plugin-plugin</artifactId>
-        <version>3.2</version>
+        <version>3.4</version>
         <configuration>
           <skipErrorNoDescriptorsFound>true</skipErrorNoDescriptorsFound>
         </configuration>

--- a/src/test/java/com/spotify/docker/BuildMojoTest.java
+++ b/src/test/java/com/spotify/docker/BuildMojoTest.java
@@ -34,9 +34,11 @@ import com.spotify.docker.client.messages.ProgressMessage;
 import org.apache.maven.execution.DefaultMavenExecutionRequest;
 import org.apache.maven.execution.MavenExecutionRequest;
 import org.apache.maven.execution.MavenSession;
+import org.apache.maven.model.Build;
 import org.apache.maven.plugin.MojoExecution;
 import org.apache.maven.plugin.MojoExecutionException;
 import org.apache.maven.plugin.testing.AbstractMojoTestCase;
+import org.apache.maven.plugin.testing.stubs.MavenProjectStub;
 import org.apache.maven.project.MavenProject;
 import org.apache.maven.project.ProjectBuilder;
 import org.apache.maven.project.ProjectBuildingRequest;
@@ -336,11 +338,12 @@ public class BuildMojoTest extends AbstractMojoTestCase {
   }
   
   private BuildMojo setupMojo(final File pom) throws Exception {
-    final MavenExecutionRequest executionRequest = new DefaultMavenExecutionRequest();
-    final ProjectBuildingRequest buildingRequest = executionRequest.getProjectBuildingRequest();
-    final ProjectBuilder projectBuilder = this.lookup(ProjectBuilder.class);
-    final MavenProject project = projectBuilder.build(pom, buildingRequest).getProject();
+    final MavenProject project = new ProjectStub(pom);
     final MavenSession session = newMavenSession(project);
+    // for some reason the superclass method newMavenSession() does not copy properties from the
+    // project model to the session. This is needed for the use of ExpressionEvaluator in BuildMojo.
+    session.getRequest().setUserProperties(project.getModel().getProperties());
+
     final MojoExecution execution = newMojoExecution("build");
     final BuildMojo mojo = (BuildMojo) this.lookupConfiguredMojo(session, execution);
     mojo.buildDirectory = "target";

--- a/src/test/java/com/spotify/docker/ProjectStub.java
+++ b/src/test/java/com/spotify/docker/ProjectStub.java
@@ -1,0 +1,51 @@
+package com.spotify.docker;
+
+import org.apache.maven.model.Build;
+import org.apache.maven.model.Model;
+import org.apache.maven.model.io.xpp3.MavenXpp3Reader;
+import org.apache.maven.plugin.testing.stubs.MavenProjectStub;
+import org.codehaus.plexus.util.ReaderFactory;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Custom stub implementation of {@link org.apache.maven.project.MavenProject}.
+ * <p>
+ * Originally taken from <a href="https://maven.apache.org/plugin-testing/maven-plugin-testing-harness/examples/complex-mojo-parameters.html">
+ *   Maven Plugin Testing Mechanism</a> guide but adapted for our use case.</p>
+ */
+public class ProjectStub extends MavenProjectStub {
+
+  public ProjectStub(File pom) {
+    MavenXpp3Reader pomReader = new MavenXpp3Reader();
+    Model model;
+    try {
+      model = pomReader.read(ReaderFactory.newXmlReader(pom));
+      setModel(model);
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+
+    setGroupId(model.getGroupId());
+    setArtifactId(model.getArtifactId());
+    setVersion(model.getVersion());
+    setName(model.getName());
+    setUrl(model.getUrl());
+    setPackaging(model.getPackaging());
+    setBuild(model.getBuild());
+
+    List<String> compileSourceRoots = new ArrayList<>();
+    compileSourceRoots.add(getBasedir() + "/src/main/java");
+    setCompileSourceRoots(compileSourceRoots);
+
+    List<String> testCompileSourceRoots = new ArrayList<>();
+    testCompileSourceRoots.add(getBasedir() + "/src/test/java");
+    setTestCompileSourceRoots(testCompileSourceRoots);
+
+    // normalize some expressions
+    getBuild().setDirectory("${project.basedir}/target");
+    getBuild().setTestOutputDirectory(new File(getBasedir(), "target/classes").getAbsolutePath());
+  }
+}

--- a/src/test/resources/pom-push-private-repo.xml
+++ b/src/test/resources/pom-push-private-repo.xml
@@ -14,14 +14,6 @@
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
   </properties>
 
-  <dependencies>
-    <dependency>
-      <groupId>com.google.guava</groupId>
-      <artifactId>guava</artifactId>
-      <version>15.0</version>
-    </dependency>
-  </dependencies>
-
   <build>
     <plugins>
       <plugin>

--- a/src/test/resources/pom-push.xml
+++ b/src/test/resources/pom-push.xml
@@ -14,14 +14,6 @@
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
   </properties>
 
-  <dependencies>
-    <dependency>
-      <groupId>com.google.guava</groupId>
-      <artifactId>guava</artifactId>
-      <version>15.0</version>
-    </dependency>
-  </dependencies>
-
   <build>
     <plugins>
       <plugin>

--- a/src/test/resources/pom-removeImage.xml
+++ b/src/test/resources/pom-removeImage.xml
@@ -14,14 +14,6 @@
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
   </properties>
 
-  <dependencies>
-    <dependency>
-      <groupId>com.google.guava</groupId>
-      <artifactId>guava</artifactId>
-      <version>15.0</version>
-    </dependency>
-  </dependencies>
-
   <build>
     <plugins>
       <plugin>

--- a/src/test/resources/pom-removeMultipleImages.xml
+++ b/src/test/resources/pom-removeMultipleImages.xml
@@ -14,14 +14,6 @@
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
   </properties>
 
-  <dependencies>
-    <dependency>
-      <groupId>com.google.guava</groupId>
-      <artifactId>guava</artifactId>
-      <version>15.0</version>
-    </dependency>
-  </dependencies>
-
   <build>
     <plugins>
       <plugin>

--- a/src/test/resources/pom-tag1.xml
+++ b/src/test/resources/pom-tag1.xml
@@ -14,14 +14,6 @@
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
   </properties>
 
-  <dependencies>
-    <dependency>
-      <groupId>com.google.guava</groupId>
-      <artifactId>guava</artifactId>
-      <version>15.0</version>
-    </dependency>
-  </dependencies>
-
   <build>
     <plugins>
       <plugin>

--- a/src/test/resources/pom-tag2.xml
+++ b/src/test/resources/pom-tag2.xml
@@ -14,14 +14,6 @@
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
   </properties>
 
-  <dependencies>
-    <dependency>
-      <groupId>com.google.guava</groupId>
-      <artifactId>guava</artifactId>
-      <version>15.0</version>
-    </dependency>
-  </dependencies>
-
   <build>
     <plugins>
       <plugin>

--- a/src/test/resources/pom-tag3.xml
+++ b/src/test/resources/pom-tag3.xml
@@ -14,14 +14,6 @@
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
   </properties>
 
-  <dependencies>
-    <dependency>
-      <groupId>com.google.guava</groupId>
-      <artifactId>guava</artifactId>
-      <version>15.0</version>
-    </dependency>
-  </dependencies>
-
   <build>
     <plugins>
       <plugin>

